[FEAT] Add Dung Beetle Optimizer (DBO) to swarm_based

ErenKayacilar opened the pull request:
This PR adds a new swarm-based optimizer: **Dung Beetle Optimizer (DBO)**.

### 🔧 What’s Included
- Implemented `OriginalDBO` in `mealpy/swarm_based/DBO.py`
- Exposed the algorithm in `mealpy/__init__.py`
- Follows the structure and coding style of existing swarm-based optimizers (e.g., COA, BFO, DMOA)
- Behavior based on the original paper:
  > Xue, J., & Shen, B. (2022). Dung beetle optimizer: A new meta-heuristic algorithm for global optimization. *The Journal of Supercomputing*, 79, 7305–7336.
- Successfully tested on the Sphere benchmark function to verify convergence and correct integration with MEALPY

<!-- 
Thanks for creating this pull request 🤗

Please make sure that the pull request is limited to one type (docs, feature, etc.) and keep it as small as possible. You can open multiple prs instead of opening a huge one.
-->

<!-- If this pull request closes an issue, please mention the issue number below -->
Closes #

---

## 📑 Description
Adds an implementation of the Dung Beetle Optimizer inspired by the original paper and integrates it into the MEALPY swarm-based module. Ensures compatibility with the current optimizer interface, population handling structure, and evaluation pipeline.

---

## ✅ Checks
- [x] My pull request adheres to the code style of this project  
- [ ] My code requires changes to the documentation  
- [ ] I have updated the documentation as required  
- [x] All local tests run successfully (basic Sphere function test)

---

## ℹ Additional Information
No breaking changes.  
The implementation is self-contained and does not modify other algorithms.

ErenKayacilar committed:
[FEAT] Add Dung Beetle Optimizer (DBO) to swarm_based

This PR adds a new swarm-based optimizer: **Dung Beetle Optimizer (DBO)**.

- Implemented `OriginalDBO` in `mealpy/swarm_based/DBO.py`
- Exposed the algorithm in `mealpy/__init__.py`
- The implementation follows the structure and coding style of existing swarm-based optimizers (e.g., COA, BFO).
- Behavior is based on the original paper:

  Xue, J., & Shen, B. (2022). Dung beetle optimizer: A new meta-heuristic algorithm for global optimization. *The Journal of Supercomputing*, 79, 7305–7336.

- Tested on a simple Sphere function (minimization) to verify convergence and correct integration with MEALPY.

thieu1995 commented on mealpy/swarm_based/DBO.py:
```diff
@@ -0,0 +1,244 @@
+                [agent.solution.copy() for agent in self.pop]
+            )
+
+        pop_array = np.array([agent.solution.copy() for agent in self.pop])
```
No need for copy() operator here. Because you don't change the pop_array, you just use it.

thieu1995 reviewed (commented):
There are many more problems I can see in the code. I will update the code for better performance.

anh9895 commented on mealpy/swarm_based/DBO.py:
```diff
@@ -0,0 +1,244 @@
+            self.pop + pop_new, self.pop_size, self.problem.minmax
+        )
+
+        # Update previous positions x(t−1) for the next iteration
```
This code should be appeared before the "Merge old and new populations". 
If not, the self._prev_positions and self.pop is always the same.

anh9895 commented on mealpy/swarm_based/DBO.py:
```diff
@@ -0,0 +1,244 @@
+        79, 7305–7336.
+    """
+
+    def __init__(
```
Write code inline, not break line like this please.

anh9895 reviewed (commented):
Please fix all the comments.

ErenKayacilar commented:
Hi @thieu1995 and @anh9895,

First of all, I sincerely apologize for the long delay in my response. I had some personal/academic commitments that kept me away from this project for a while.

Thank you so much for the detailed feedback and code reviews. I've carefully reviewed all your comments regarding memory optimization, code structure, and the positioning of the population updates.

I am still very much interested in contributing the Dung Beetle Optimizer to Mealpy. I will start implementing the requested changes and resolve the merge conflicts as soon as possible.

Thanks for your patience

ErenKayacilar committed:
Merge branch 'master' into feature/dung-beetle-optimizer

ErenKayacilar committed:
Refactor DBO.py based on review feedback

Removed custom generate_empty_agent() override since DBO has no additional agent properties; parent class method is sufficient.

Removed unused Agent import.

Removed unnecessary _prev_positions size check in 
evolve()  — population size is always consistent.

Removed redundant copy() calls on pop_array and g_best/g_worst since they are only read, not mutated.

Eliminated extra prev_array variable; using self._prev_positions directly to avoid wasted memory.

Replaced manual best/worst agent logic with built-in self.get_worst_agent() from the Optimizer base class.

Moved _prev_positions update to before the merge step so it correctly captures the current population state, not the already-merged one.

Reformatted init
 signature to inline style as per code style guidelines.

ErenKayacilar committed:
Update __init__.py
